### ISSUE TITLE
failed to list *v1.Pod: proto: wrong wireType = 0 for field NamespaceSelector

### ISSUE THREAD

**Issue by alexandnpu** [prometheus/prometheus#13878]
### What did you do?

We have an old kubernetes cluster, version 1.11.3.

Today I set up one prometheus, using v2.49.1, to get some info from it, buf failed. 

then I tried some other versions of prometheus, and found that starting from v2.40, all the versions failed.
but v2.27.8 is ok.

In the [release note of v2.40](https://github.com/prometheus/prometheus/releases/tag/v2.40.0), I see there is an enhancement, [Kubernetes SD: Use protobuf encoding](https://github.com/prometheus/prometheus/pull/11353).

There, it hard codes the content type to "application/vnd.kubernetes.protobuf".

Can we make an configuration option to do so ? to make it compatible with old kubernetes cluster?

### What did you expect to see?

Prometheus should be compatible with old kubernetes clusters.
That is, newer prometheus should be able to monitor old kubernetes clusters with no error.
if no default way to work with old kubernetes, there should be some option to make it work.

As prometheus is easy to upgrade, but kubernetes is not.

### What did you see instead? Under which circumstances?

prometheus failed to do kubernetes service discovery

### Prometheus configuration file

```yaml
- job_name: 'my-k8s'
    scrape_interval: 120s
    scrape_timeout: 90s

    kubernetes_sd_configs:
    - role: endpoints
      api_server: https://xxxxxx:443
      tls_config:
        cert_file: aaaaaaaa
        key_file: bbbbbbb
        insecure_skip_verify: true
```

### Logs

```text
below is the log I got from version 2.40.7.

> ts=2024-04-02T08:10:13.039Z caller=klog.go:108 level=warn component=k8s_client_runtime func=Warningf msg="pkg/mod/k8s.io/client-go@v0.25.3/tools/cache/reflector.go:169: failed to list *v1.Pod: proto: wrong wireType = 0 for field NamespaceSelector"
ts=2024-04-02T08:10:13.040Z caller=klog.go:116 level=error component=k8s_client_runtime func=ErrorDepth msg="pkg/mod/k8s.io/client-go@v0.25.3/tools/cache/reflector.go:169: Failed to watch *v1.Pod: failed to list *v1.Pod: proto: wrong wireType = 0 for field NamespaceSelector"
```

**Comment by alexandnpu** [comment]
Today I come across the [protobuf introduction document in kubernetes](https://github.com/kubernetes/design-proposals-archive/blob/main/api-machinery/protobuf.md#optimistically-attempt-to-send-and-receive-requests-using-protobuf)

> Using multiple Accept values:

> Accept: application/vnd.kubernetes.protobuf, application/json

> clients can indicate their preferences and handle the returned Content-Type using whatever the server responds. On 
 update operations, clients can try protobuf and if they receive a 415 error, record that and fall back to JSON. Allows the client to be backwards compatible with any server, but comes at the cost of some implementation complexity.

should we try using protobuf first; if fails, we fall back to json?

**Comment by alexandnpu** [comment]
> protobuf was added since Kube 1.3 I think,1.11.3 should be able to handle it.
Are you deploying a tweaked/custom kube API?

I am not quite sure. I will make sure that with our k8s team.


> Unfortunately, I don't have any `1.11` Kube cluster around. However, you could try compiling and deploying Prometheus after adding `kcfg.AcceptContentTypes = "application/vnd.kubernetes.protobuf,application/json"` to see if that allows the client to fall back to application/json in our case. If this works, we could implement the change.
> 

below are the versions that I tried. I downloaded them from the releases pages.
```
prometheus-2.37.8.linux-amd64
prometheus-2.40.7.linux-amd64
prometheus-2.45.1.linux-amd64
prometheus-2.49.1.linux-amd64
prometheus-2.50.1.linux-amd64
prometheus-2.51.1.linux-amd64
```

`prometheus-2.37.8.linux-amd64` is the version that just before adding `kcfg.AcceptContentTypes = "application/vnd.kubernetes.protobuf,application/json"`. 

`prometheus-2.37.8.linux-amd64` works fine. (sorry that I had a typo in the original post, I typed it as `2.27.8`)
but all the versions afterwards fail to list pods.

**Comment by machine424** [comment]
Thanks for trying that out.
(Just to be sure we're on the same page, you mean `kcfg.ContentType = "application/vnd.kubernetes.protobuf"` and not `kcfg.AcceptContentTypes = "application/vnd.kubernetes.protobuf"`?)

Let's see if the other maintainers could help with:

> I don't know what changed between https://github.com/prometheus/prometheus/pull/6802 (first attempt to switch to protobuf by default) and https://github.com/prometheus/prometheus/pull/11353 (where the switch was done). Maybe we implicitly decided that older Kube versions are no longer supported. I'm not aware of any compatibility policy regarding Kube (maybe we follow client-go's?). Maybe @roidelapluie @brancz could provide some clarification on this matter.

**Comment by alexandnpu** [comment]
> Thanks for trying that out. (Just to be sure we're on the same page, you mean `kcfg.ContentType = "application/vnd.kubernetes.protobuf"` and not `kcfg.AcceptContentTypes = "application/vnd.kubernetes.protobuf"`?) It seems to be an incompatibility between the client used by prom and the kube API.
> 

yes, I tried `kcfg.ContentType = "application/vnd.kubernetes.protobuf"`, and left `kcfg.AcceptContentTypes` empty.

Sorry about the confusion, I just copied your earlier reply, and didn't noticed the difference.  😂
